File: exulanica/grammar/grammars/city/massing.py
"""Massing: a building's tiers, storeys and roof, and what stands on the roof. Record shapes only.

**Storeys.** Storey 0 is the ground storey, ``ground_storey_height_mm`` tall from the building's
``base_elevation_mm``; every storey above it is ``upper_storey_height_mm`` tall. Storey ``k``'s
floor is at ``base + ground`` plus ``(k - 1) * upper`` for ``k >= 1``.

**Tiers are the setbacks, written out.** Tier 0's ring is the footprint and starts at storey 0.
Each tier covers a contiguous run of storeys, the next tier starts on the storey after, and the
last ends on the top storey. A tier's ring lies within the ring below it, touching it where the
walls run flush (a party wall usually does), so a setback is a ring the generator computed, never
an offset a reader has to derive. Where a tier is set back, the exposed part of the tier below is
a flat roof terrace with that tier's parapet. Light wells are rings strictly inside their tier;
their walls take the building's own ``wall`` material and carry no openings.

**The roof is a geometric form, not a catalog word.** ``roof_family`` is a label.
``roof_form`` decides the shape: ``flat`` is a deck at the top tier's wall top with the top tier's
parapet; ``ridge`` needs a convex four-sided top tier and ``ridge_mm`` holds the ridge's two plan
points, which are the floored midpoints of two opposite edges (the gable edges). The ridge stands
``roof_rise_mm`` above the wall top; the two gable edges carry vertical triangles, and the other
two edges each carry one plane up to the ridge. A flat roof has no ridge and no rise.

**Rooftop objects** stand only on flat roofs, on a roof region that is not covered by a tier
above, with their base on that region's deck. Their shapes are explicit parts.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import ClassVar, Final

from exulanica.grammar import shapes
from exulanica.grammar.errors import InvalidRecordError
from exulanica.grammar.geometry import (
    Extent,
    extent_contains_ring,
    ring_inside_ring,
    ring_is_convex,
    ring_within_ring,
    rings_disjoint,
)
from exulanica.grammar.grammars.city._skeleton import skeleton
from exulanica.grammar.grammars.city.common import (
    FORM_PART_SHAPE,
    FormPart,
    direction_fields,
    extent_field,
    require_direction,
    require_point_in_extent,
)

__all__ = [
    "MASSING_SHAPE",
    "ROOFTOP_SHAPE",
    "ROOF_FORMS",
    "STAGE",
    "STAGE_ID",
    "STAGE_VERSION",
    "TIER_SHAPE",
    "MassingRecord",
    "RooftopObjectRecord",
    "Tier",
    "storey_floor_mm",
    "tier_top_mm",
]

STAGE_ID: Final = "massing"
STAGE_VERSION: Final = 2
ROOF_FORMS: Final = ("flat", "ridge")


@dataclass(frozen=True, slots=True)
class Tier:
    first_storey: int
    last_storey: int
    ring_mm: tuple[tuple[int, int], ...]
    light_wells_mm: tuple[tuple[tuple[int, int], ...], ...]
    #: The parapet round this tier's roof or terrace; 0 is none.
    parapet_height_mm: int


def _tier_rules(tier: Tier) -> None:
    if tier.first_storey > tier.last_storey:
        raise InvalidRecordError("a tier's first storey is at most its last")
    for index, well in enumerate(tier.light_wells_mm):
        if not ring_inside_ring(well, tier.ring_mm):
            raise InvalidRecordError(f"light well {index} lies strictly inside its tier")
        for other in tier.light_wells_mm[index + 1 :]:
            if not rings_disjoint(well, other):
                raise InvalidRecordError("light wells do not meet")


TIER_SHAPE: Final = shapes.RecordShape(
    Tier,
    (
        shapes.integer("first_storey", 0),
        shapes.integer("last_storey", 0),
        shapes.ring("ring_mm"),
        shapes.rings("light_wells_mm"),
        shapes.integer("parapet_height_mm", 0, 2_000),
    ),
    rules=(shapes.RecordRule("tier_storeys_and_wells", _tier_rules),),
)


@dataclass(frozen=True, slots=True)
class MassingRecord:
    """The building. Its identity is the building's, which every part of it refers to."""

    RECORD_KIND: ClassVar[str] = "city.massing"
    RECORD_VERSION: ClassVar[int] = 2

    identity: str
    parcel_identity: str
    building_ordinal: int
    base_elevation_mm: int
    #: Keys into the typology and era catalogs. Labels.
    typology: str
    era: str
    storeys: int
    ground_storey_height_mm: int
    upper_storey_height_mm: int
    tiers: tuple[Tier, ...]
    #: A key into the roof-family catalog. A label: ``roof_form`` is the geometry.
    roof_family: str
    roof_form: str
    ridge_mm: tuple[tuple[int, int], ...]
    roof_rise_mm: int
    extent: Extent


def storey_floor_mm(record: MassingRecord, storey: int) -> int:
    """The floor level of ``storey``, in millimetres above the datum."""
    if storey == 0:
        return record.base_elevation_mm
    return (
        record.base_elevation_mm
        + record.ground_storey_height_mm
        + (storey - 1) * record.upper_storey_height_mm
    )


def tier_top_mm(record: MassingRecord, tier: Tier) -> int:
    """The wall top of a tier: the ceiling of its last storey."""
    return storey_floor_mm(record, tier.last_storey + 1)


def _floored_midpoint(a: tuple[int, int], b: tuple[int, int]) -> tuple[int, int]:
    return (a[0] + b[0]) // 2, (a[1] + b[1]) // 2


def _massing_rules(record: MassingRecord) -> None:
    expected_first = 0
    for index, tier in enumerate(record.tiers):
        if tier.first_storey != expected_first:
            raise InvalidRecordError(f"tier {index} starts on storey {expected_first}")
        expected_first = tier.last_storey + 1
        if index and not ring_within_ring(tier.ring_mm, record.tiers[index - 1].ring_mm):
            raise InvalidRecordError(f"tier {index} lies within the tier below it")
    if expected_first != record.storeys:
        raise InvalidRecordError("the tiers cover every storey exactly once")
    top = record.tiers[-1]
    if record.roof_form == "flat":
        if record.ridge_mm or record.roof_rise_mm:
            raise InvalidRecordError("a flat roof has no ridge and no rise")
    else:
        ring = top.ring_mm
        if len(ring) != 4 or not ring_is_convex(ring):
            raise InvalidRecordError("a ridge roof stands on a convex four-sided top tier")
        if len(record.ridge_mm) != 2 or record.roof_rise_mm < 1:
            raise InvalidRecordError("a ridge roof states its two ridge points and its rise")
        pairs = {
            (
                _floored_midpoint(ring[k], ring[k + 1]),
                _floored_midpoint(ring[k + 2], ring[(k + 3) % 4]),
            )
            for k in (0, 1)
        }
        if tuple(record.ridge_mm) not in pairs:
            raise InvalidRecordError(
                "a ridge runs between the floored midpoints of edges k and k + 2, in that order"
            )
    if not extent_contains_ring(record.extent, record.tiers[0].ring_mm):
        raise InvalidRecordError("a building's footprint lies inside its extent")
    summit = tier_top_mm(record, top) + max(record.roof_rise_mm, top.parapet_height_mm)
    if record.extent.min_z_mm > record.base_elevation_mm or record.extent.max_z_mm < summit:
        raise InvalidRecordError(f"a building's extent spans its base to {summit}")


MASSING_SHAPE: Final = shapes.RecordShape(
    MassingRecord,
    (
        shapes.identity("identity"),
        shapes.identity("parcel_identity", "city.parcel"),
        shapes.integer("building_ordinal", 0),
        shapes.integer("base_elevation_mm"),
        shapes.key("typology", "typology"),
        shapes.key("era", "era"),
        shapes.integer("storeys", 1, 120),
        shapes.integer("ground_storey_height_mm", 4_000, 6_000),
        shapes.integer("upper_storey_height_mm", 2_400, 6_000),
        shapes.records("tiers", TIER_SHAPE, count_minimum=1),
        shapes.key("roof_family", "roof-family"),
        shapes.choice("roof_form", ROOF_FORMS),
        shapes.points("ridge_mm", 2, count_minimum=0, count_maximum=2),
        shapes.integer("roof_rise_mm", 0, 20_000),
        extent_field(),
    ),
    rules=(shapes.RecordRule("massing_tiers_and_roof", _massing_rules),),
    identity=shapes.IdentityRule(
        "building", owner_field="parcel_identity", ordinal_field="building_ordinal"
    ),
    extent_field="extent",
)


@dataclass(frozen=True, slots=True)
class RooftopObjectRecord:
    """Plant, a tank, a lift overrun: explicit parts at a point on a flat roof."""

    RECORD_KIND: ClassVar[str] = "city.rooftop_object"
    RECORD_VERSION: ClassVar[int] = 1

    identity: str
    building_identity: str
    object_ordinal: int
    #: A key into the rooftop-object catalog. A label: ``parts`` are the geometry.
    object_class: str
    x_mm: int
    y_mm: int
    z_mm: int
    facing_dx_mm: int
    facing_dy_mm: int
    parts: tuple[FormPart, ...]
    #: The least distance kept from every roof edge and parapet.
    clearance_mm: int
    extent: Extent


def _rooftop_rules(record: RooftopObjectRecord) -> None:
    require_direction("facing", record.facing_dx_mm, record.facing_dy_mm)
    require_point_in_extent(
        "the object's base", record.extent, record.x_mm, record.y_mm, record.z_mm
    )


ROOFTOP_SHAPE: Final = shapes.RecordShape(
    RooftopObjectRecord,
    (
        shapes.identity("identity"),
        shapes.identity("building_identity", "city.massing"),
        shapes.integer("object_ordinal", 0),
        shapes.key("object_class", "rooftop-object"),
        shapes.integer("x_mm"),
        shapes.integer("y_mm"),
        shapes.integer("z_mm"),
        *direction_fields(),
        shapes.records("parts", FORM_PART_SHAPE, count_minimum=1),
        shapes.integer("clearance_mm", 0, 10_000),
        extent_field(),
    ),
    rules=(shapes.RecordRule("rooftop_object_placed", _rooftop_rules),),
    identity=shapes.IdentityRule(
        "rooftop_object", owner_field="building_identity", ordinal_field="object_ordinal"
    ),
    extent_field="extent",
)

STAGE: Final = skeleton(STAGE_ID, STAGE_VERSION, MASSING_SHAPE, ROOFTOP_SHAPE)

File: web/packages/loom-tess/src/core/massing.ts
/**
 * THE MASSING RULE: A BUILDING'S OWN FACES, FROM ITS TIERS AND ITS ROOF, IN INTEGERS.
 *
 * A massing states everything this needs and leaves nothing to be taken by eye: storey heights from
 * a base elevation, tiers as rings with the storeys each covers, light wells as rings inside a
 * tier, a parapet height per tier, and a roof that is either a deck or a ridge between two stated
 * points. Every face here is one of those facts turned into triangles.
 *
 *   1. WALLS. Each tier's ring edge, from the floor of its first storey to the ceiling of its last.
 *      A tier stacks on the one below, so no wall crosses a storey its tier does not cover. A light
 *      well's walls run the same way, facing inward.
 *   2. DECKS. The top tier's ring is its roof, and a lower tier's ring less the ring standing on it
 *      is a terrace. Both are cut by the shared piece carve, so a terrace is what is exactly left of
 *      the tier rather than an offset taken from the setback. Light wells are cut out the same way.
 *   3. PARAPETS. A parapet stands where a deck meets the open air: along the parts of a tier's ring
 *      edges that no ring stands on, taken as intervals along each edge, in integers.
 *   4. A RIDGE ROOF. Two planes from the eave edges up to the stated ridge, and a gable on each of
 *      the two edges the ridge runs between. The grammar holds a ridge roof to a convex four-sided
 *      top tier, so each plane is one quad and needs no further cutting.
 *
 * Surface frames are the grammar's (`common`): a vertical face runs `s` along its own ring edge
 * from that edge's start with `t = base_elevation_mm - z`, and a roof, a pitched plane included,
 * takes `s = x` and `t = y` in plan.
 *
 * WHAT THIS DOES NOT DRAW, AND WHY IT DRAWS THE WALL TODAY. A facade is laid out on a tier edge and
 * covers part of it. Until a facade DRAWS, the building's own wall is what stands there, exactly as
 * terrain draws the ground until something drawn covers it. When the facade rule lands, a wall
 * yields to the facade surfaces that are drawn, by that same rule, and never to a facade record
 * that only states it will be there. Nothing is drawn twice in one place at any point.
 */
import { add, exact, measureAgainst, multiply, subtract } from './integer-math.js';
import type { Plan } from './integer-math.js';
import { faceOn, HORIZONTAL, piecesOf, runOf, Surface, VERTICAL } from './faces.js';
import type { FaceCover } from './faces.js';
import type { Piece } from './pieces.js';
import { leftTurning, openRegions } from './piece-carve.js';
import { requireSimpleRing, triangulateRing } from './ring-triangulation.js';

/** A tier of a building: the storeys it covers, its ring, its light wells and its parapet. */
export interface Tier {
  readonly first_storey: number;
  readonly last_storey: number;
  readonly ring_mm: readonly Plan[];
  readonly light_wells_mm: readonly (readonly Plan[])[];
  readonly parapet_height_mm: number;
}

/** What this rule reads of a massing record. */
export interface MassingFields {
  readonly base_elevation_mm: number;
  readonly ground_storey_height_mm: number;
  readonly upper_storey_height_mm: number;
  readonly storeys: number;
  readonly tiers: readonly Tier[];
  readonly roof_form: string;
  readonly ridge_mm: readonly Plan[];
  readonly roof_rise_mm: number;
}

/** The roof form that rises to a ridge; every other form this version draws is a deck. */
const RIDGE = 'ridge';

/** The roles a building's own faces take, each owned by the massing kind in the grammar. */
const WALL = 'wall';
const ROOF = 'roof';
const PARAPET = 'parapet';

/** `exulanica.grammar.grammars.city.massing.storey_floor_mm`. */
export function storeyFloorMm(fields: MassingFields, storey: number, where: string): number {
  if (storey === 0) return exact(fields.base_elevation_mm, where);
  return add(
    add(fields.base_elevation_mm, fields.ground_storey_height_mm, where),
    multiply(subtract(storey, 1, where), fields.upper_storey_height_mm, where),
    where,
  );
}

/** `exulanica.grammar.grammars.city.massing.tier_top_mm`: the ceiling of a tier's last storey. */
export function tierTopMm(fields: MassingFields, tier: Tier, where: string): number {
  return storeyFloorMm(fields, add(tier.last_storey, 1, where), where);
}

/** Whether a point lies on the closed segment from `from` to `to`, exactly. */
function onSegment(from: Plan, to: Plan, point: Plan, where: string): boolean {
  const side = subtract(
    multiply(subtract(to[0], from[0], where), subtract(point[1], from[1], where), where),
    multiply(subtract(to[1], from[1], where), subtract(point[0], from[0], where), where),
    where,
  );
  if (side !== 0) return false;
  const within = (low: number, high: number, value: number): boolean => {
    if (low <= high) return low <= value && value <= high;
    return high <= value && value <= low;
  };
  return within(from[0], to[0], point[0]) && within(from[1], to[1], point[1]);
}

/**
 * The parts of one ring edge that no ring standing on the tier covers, as intervals along the edge.
 * A ring above meets the edge along whole sub-segments, both rings being the generator's own
 * integers, so subtracting intervals is exact.
 */
function openAlong(from: Plan, to: Plan, standing: readonly (readonly Plan[])[], where: string): [number, number][] {
  const direction: Plan = [subtract(to[0], from[0], where), subtract(to[1], from[1], where)];
  const run = measureAgainst(from, direction, to, where).along;
  const covered: [number, number][] = [];
  for (const ring of standing) {
    ring.forEach((point, index) => {
      const next = ring[(index + 1) % ring.length]!;
      if (!onSegment(from, to, point, where)) return;
      if (!onSegment(from, to, next, where)) return;
      const low = measureAgainst(from, direction, point, where).along;
      const high = measureAgainst(from, direction, next, where).along;
      covered.push(low <= high ? [low, high] : [high, low]);
    });
  }
  covered.sort((first, second) => first[0] - second[0]);
  const open: [number, number][] = [];
  let at = 0;
  for (const [low, high] of covered) {
    if (low > at) open.push([at, low]);
    if (high > at) at = high;
  }
  if (at < run) open.push([at, run]);
  return open;
}

/**
 * A ridge roof on a convex four-sided top tier. The ridge runs between the floored midpoints of two
 * opposite edges, `roof_rise_mm` above the wall top: those two edges carry gables and the other two
 * each carry one plane, drawn as two triangles since four corners that far apart need not be
 * coplanar.
 */
function ridgeRoof(
  fields: MassingFields,
  ring: readonly Plan[],
  top: number,
  datum: number,
  roof: Surface,
  walls: Surface,
  where: string,
): void {
  const [low, high] = [fields.ridge_mm[0]!, fields.ridge_mm[1]!];
  const crown = add(top, fields.roof_rise_mm, where);
  const at = ring.findIndex((point, index) => onSegment(point, ring[(index + 1) % ring.length]!, low, where));
  if (at < 0) throw new RangeError(`${where}: a ridge that starts on no edge of its top tier`);
  const corner = (step: number): Plan => ring[(at + step) % ring.length]!;
  // The eave edges: each runs from the gable holding one ridge end to the gable holding the other.
  for (const [from, to, near, far] of [
    [corner(1), corner(2), high, low],
    [corner(3), corner(0), low, high],
  ] as [Plan, Plan, Plan, Plan][]) {
    roof.face(
      roof.corner(from, top, from[0], from[1]),
      roof.corner(to, top, to[0], to[1]),
      roof.corner(near, crown, near[0], near[1]),
      roof.corner(far, crown, far[0], far[1]),
    );
  }
  for (const [from, to, point] of [
    [corner(0), corner(1), low],
    [corner(2), corner(3), high],
  ] as [Plan, Plan, Plan][]) {
    const direction: Plan = [subtract(to[0], from[0], where), subtract(to[1], from[1], where)];
    const run = measureAgainst(from, direction, to, where).along;
    const along = measureAgainst(from, direction, point, where).along;
    walls.face(
      walls.corner(from, top, 0, subtract(datum, top, where)),
      walls.corner(to, top, run, subtract(datum, top, where)),
      walls.corner(point, crown, along, subtract(datum, crown, where)),
    );
  }
}

/**
 * The parts of one tier edge's wall that no facade draws over, as rectangles in the edge's frame.
 * The boundaries of every cover, and the wall's own, cut the wall into cells; a cell no cover holds
 * is drawn. Cells that meet share their whole edge, so nothing cracks between them.
 */
function openWall(
  run: number,
  base: number,
  top: number,
  covers: readonly FaceCover[],
): [number, number, number, number][] {
  if (covers.length === 0) return [[0, run, base, top]];
  const cuts = (low: number, high: number, values: readonly number[]): number[] => {
    const inside = values.filter((value) => value > low && value < high);
    return [low, ...inside.sort((first, second) => first - second), high]
      .filter((value, index, all) => index === 0 ? true : value !== all[index - 1]);
  };
  const along = cuts(0, run, covers.flatMap((cover) => [cover.low, cover.high]));
  const up = cuts(base, top, covers.flatMap((cover) => [cover.base, cover.top]));
  const open: [number, number, number, number][] = [];
  for (let u = 0; u + 1 < along.length; u += 1) {
    for (let z = 0; z + 1 < up.length; z += 1) {
      const cell: [number, number, number, number] = [along[u]!, along[u + 1]!, up[z]!, up[z + 1]!];
      const held = covers.some((cover) =>
        cover.low <= cell[0] && cell[1] <= cover.high && cover.base <= cell[2] && cell[3] <= cover.top);
      if (!held) open.push(cell);
    }
  }
  return open;
}

/**
 * A building's own faces, by the rule above: its walls, its roofs and terraces, and its parapets.
 * `covered` names the parts of its tier edges that facades draw over, which its own walls give up.
 */
export function massingPieces(fields: MassingFields, covered: readonly FaceCover[], where: string): Piece[] {
  const walls = new Surface(WALL, VERTICAL);
  const roof = new Surface(ROOF, HORIZONTAL);
  const parapets = new Surface(PARAPET, VERTICAL);
  const datum = fields.base_elevation_mm;
  fields.tiers.forEach((tier, index) => {
    const ring = leftTurning(requireSimpleRing(tier.ring_mm, where), where);
    const base = storeyFloorMm(fields, tier.first_storey, where);
    const top = tierTopMm(fields, tier, where);
    const above = fields.tiers[index + 1];
    const standing: (readonly Plan[])[] = [];
    if (above !== undefined) standing.push(leftTurning(requireSimpleRing(above.ring_mm, where), where));

    ring.forEach((point, corner) => {
      const next = ring[(corner + 1) % ring.length]!;
      const run = runOf(point, next, where);
      const covers = covered.filter((cover) => cover.tier === index && cover.edge === corner);
      for (const [low, high, from, to] of openWall(run, base, top, covers)) {
        faceOn(walls, { from: point, to: next, run }, low, high, from, to, 0, datum, where);
      }
    });
    for (const well of tier.light_wells_mm) {
      const turned = leftTurning(requireSimpleRing(well, where), where);
      // A well's walls face into the well, so its walk is taken the other way round.
      const inward = [...turned].reverse();
      inward.forEach((point, corner) => {
        const next = inward[(corner + 1) % inward.length]!;
        const run = runOf(point, next, where);
        faceOn(walls, { from: point, to: next, run }, 0, run, base, top, 0, datum, where);
      });
      standing.push(turned);
    }

    const ridged = fields.roof_form === RIDGE && index === subtract(fields.tiers.length, 1, where);
    if (ridged) {
      ridgeRoof(fields, ring, top, datum, roof, walls, where);
    } else {
      for (const walk of openRegions(ring, standing, where)) {
        const corners = walk.map((point) => roof.corner(point, top, point[0], point[1]));
        for (const index of triangulateRing(walk, where)) roof.triangles.push(corners[index]!);
      }
    }

    if (tier.parapet_height_mm > 0) {
      const crown = add(top, tier.parapet_height_mm, where);
      ring.forEach((point, corner) => {
        const next = ring[(corner + 1) % ring.length]!;
        const run = runOf(point, next, where);
        for (const [low, high] of openAlong(point, next, standing, where)) {
          faceOn(parapets, { from: point, to: next, run }, low, high, top, crown, 0, datum, where);
        }
      });
    }
  });
  return piecesOf([walls, roof, parapets]);
}
